gadget/update.go

```diff
@@ -1437,21 +1450,30 @@ func arePossibleOffsetsCompatible(vss1 []VolumeStructure, idx1 int, vss2 []Volum
 		minStructureOffset(vss1, idx1) <= maxStructureOffset(vss2, idx2)
 }
 
-func canUpdateStructure(fromVss []VolumeStructure, fromIdx int, toVss []VolumeStructure, toIdx int, schema string) error {
-	from := &fromVss[fromIdx]
-	to := &toVss[toIdx]
-	if schema == schemaGPT && from.Name != to.Name {
+// canUpdateStructure checks gadget compatibility on updates, looking only at
+// features that are not reflected on the installed disk (for this we check
+// elsewhere the new gadget against the actual disk content).
+//
+// Partial properties are not checked as they will be checked against the real
+// disk later, in EnsureVolumeCompatibility. TODO Some checks should maybe
+// happen only there even for non-partial gadgets.
+func canUpdateStructure(fromV *Volume, fromIdx int, toV *Volume, toIdx int) error {
+	from := &fromV.Structure[fromIdx]
+	to := &toV.Structure[toIdx]
+	if !toV.HasPartial(PartialSchema) && toV.Schema == schemaGPT && from.Name != to.Name {
 		// partition names are only effective when GPT is used
 		return fmt.Errorf("cannot change structure name from %q to %q",
 			from.Name, to.Name)
 	}
-	if !arePossibleSizesCompatible(from, to) {
-		return fmt.Errorf("new valid structure size range [%v, %v] is not compatible with current ([%v, %v])",
-			to.MinSize, to.Size, from.MinSize, from.Size)
-	}
-	if !arePossibleOffsetsCompatible(fromVss, fromIdx, toVss, toIdx) {
-		return fmt.Errorf("new valid structure offset range [%v, %v] is not compatible with current ([%v, %v])",
-			minStructureOffset(toVss, toIdx), maxStructureOffset(toVss, toIdx), minStructureOffset(fromVss, fromIdx), maxStructureOffset(fromVss, fromIdx))
+	if !toV.HasPartial(PartialSize) {
```

Comment:
shouldn't we actually check if the sides actually set a size as well? and the sizes if they are even if in general they are partial and could be omitted

Reply:
Right, this is wrong as it is checking the volume, not the structure. Fixed and unit tests added.

Reply:
Right, this is wrong as it is checking the volume, not the structure. Fixed and unit tests added.